src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1207,28 +1213,126 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
   8. Need to figure out a way to get all versions to be deleted in case of versioned objects when a version is not specified.
   9. If the object is not in object directory call next->delete_obj
   */
-  rgw::d4n::CacheObj obj = rgw::d4n::CacheObj{ // TODO: Add logic to ObjectDirectory del method to also delete all blocks belonging to that object
-			     .objName = source->get_key().get_oid(),
-			     .bucketName = source->get_bucket()->get_name()
-			   };
 
-  int ret = -1;
-  if ((ret = source->driver->get_obj_dir()->del(dpp, &obj, y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): ObjectDirectory del() method failed, ret=" << ret << dendl;
+  rgw::sal::Attrs attrs;
+  std::string head_oid_in_cache;
+  if (!source->check_head_exists_in_cache_get_oid(dpp, head_oid_in_cache, attrs, y)) {
+    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): calling next delete_obj" << dendl;
+    return next->delete_obj(dpp, y, flags);
+  } else {
+    rgw::d4n::CacheObj obj = rgw::d4n::CacheObj{ 
+			       .objName = source->get_name(),
+			       .bucketName = source->get_bucket()->get_name()
+			     };
 
-  Attrs::iterator attrs;
-  Attrs currentattrs = source->get_attrs();
-  std::vector<std::string> currentFields;
-  
-  /* Extract fields from current attrs */
-  for (attrs = currentattrs.begin(); attrs != currentattrs.end(); ++attrs) {
-    currentFields.push_back(attrs->first);
-  }
+    int ret = -1;
+    auto rgwObj = source->get_obj();
+    ret = source->get_obj_attrs(y, dpp, &rgwObj);
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): failed to get obj attrs, ret=" << ret << dendl;
+      return next->delete_obj(dpp, y, flags);
+    }
+    
+    std::string version, prefix, policy_prefix;
+    if (source->have_instance()) {
+      version = source->get_instance();
+    } else {
+      version = source->get_object_version();
+    }
+    
+    prefix = source->get_bucket()->get_name() + "_" + version + "_" + obj.objName;
+    policy_prefix = prefix;
+
+    auto blockDir = source->driver->get_block_dir(); 
+    rgw::d4n::CacheBlock block;
+    block.cacheObj.objName = source->get_name();
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    block.blockID = 0;
+    block.size = 0;
+
+    // retrieve head object from directory to check for dirty flag
+    if ((ret = blockDir->get(dpp, &block, y)) < 0) { // Sam: can we somehow return the block from the check_head_exists_in_cache_get_oid method to avoid this redis call?
```

Comment:
Like check_head_exists_in_cache_get_oid() returns attrs, it can return the `dirty` flag as well, if only the `dirty` flag is needed. also head_oid_in_cache, returns the oid of the head in cache, so there is no need for it to be derived in line 1259 etc.

Reply:
You mean the `dirty` flag gets returned as part of the `head_oid_in_cache` string if the method returns `true`, correct?